src/util/expr.cpp

```diff
@@ -539,3 +539,184 @@ const_unique_depth_iteratort exprt::unique_depth_cbegin() const
 { return const_unique_depth_iteratort(*this); }
 const_unique_depth_iteratort exprt::unique_depth_cend() const
 { return const_unique_depth_iteratort(); }
+
+/// We use the precendences that most readers expect
+/// (i.e., the ones you learn in primary school),
+/// and stay clear of the surprising ones that C has.
+static bool precedence(const exprt &a, const exprt &b)
+{
+  if((a.id()==ID_mult || a.id()==ID_div) &&
+     (b.id()==ID_plus || b.id()==ID_minus))
+    return true;
+
+  if((a.id()==ID_equal || a.id()==ID_notequal ||
+      a.id()==ID_lt || a.id()==ID_gt ||
+      a.id()==ID_le || a.id()==ID_ge) &&
+     (b.id()==ID_and || b.id()==ID_or))
+    return true;
+
+  return false;
+}
+
+/// This turns a multi-ary expression into a string,
+/// adding parentheses where indicated by the precedence.
+static std::string as_string_multi_ary(const exprt &src)
+{
+  bool first=true;
+
+  std::string result;
+
+  for(const auto &op : src.operands())
+  {
+    if(first)
+      first=false;
+    else
+    {
+      result+=' ';
+      result+=id2string(src.id());
+      result+=' ';
+    }
+
+    const bool need_parentheses=
+      !op.operands().empty() &&
+      !precedence(op, src);
+
+    if(need_parentheses)
+      result+='(';
+
+    result+=op.as_string();
+
+    if(need_parentheses)
+      result+=')';
+  }
+
+  return result;
+}
+
+/// This turns a binary expression into a string,
+/// adding parentheses where indicated by the precedence.
+static std::string as_string_binary(const exprt &src)
+{
+  if(src.operands().size()!=2)
+    return src.pretty();
+  else
+    return as_string_multi_ary(src);
+}
+
+/// This turns a unary expression into a string,
+/// adding parentheses very aggressively.
+static std::string as_string_unary(const exprt &src)
+{
+  if(src.operands().size()!=1)
+    return src.pretty();
+
+  std::string result;
+
+  if(src.id()==ID_not)
+    result="!";
+  else if(src.id()==ID_unary_minus)
+    result="-";
+  else
+    return src.pretty();
+
+  if(src.op0().has_operands())
+    return result+'('+src.op0().as_string()+')';
+  else
+    return result+src.op0().as_string();
+}
+
+static std::string as_string(const constant_exprt &src)
+{
+  auto type=src.type().id();
+
+  if(type==ID_bool)
+  {
+    if(src.is_true())
+      return "true";
+    else if(src.is_false())
+      return "false";
+    else
+      return src.pretty();
+  }
+  else if(type==ID_unsignedbv || type==ID_signedbv)
+    return integer2string(*numeric_cast<mp_integer>(src));
+  else if(type==ID_integer)
+    return id2string(src.get_value());
+  else if(type==ID_floatbv)
+    return ieee_floatt(src).to_ansi_c_string();
+  else
+    return src.pretty();
+}
+
+// The below generates a string in a generic syntax
+// that is inspired by C/C++/Java, and is meant for debugging
+// purposes.
+std::string exprt::as_string() const
+{
+  if(id()==ID_plus || id()==ID_mult ||
+     id()==ID_and || id()==ID_or)
+    return as_string_multi_ary(*this);
```

Comment:
See below for why this should be a non-member function. Either way, it should be `return as_string(to_multi_ary_expr(...));`